server/api/src/rsGeneralAdmin.cpp

```diff
@@ -25,8 +27,142 @@
+
+    // build genquery to find active or stale "rebalance operation" entries for this resource
+    genQueryOut_t* gen_out = nullptr;
+    int status = 0;
+    char tmp_str[MAX_NAME_LEN];
+    genQueryInp_t gen_inp;
+    memset( &gen_inp, 0, sizeof( genQueryInp_t ) );
+
+    snprintf( tmp_str, MAX_NAME_LEN, "='%s'", _resource_name.c_str() );
+    addInxVal( &gen_inp.sqlCondInp, COL_D_RESC_NAME, tmp_str );
+    snprintf( tmp_str, MAX_NAME_LEN, "='rebalance_operation'" );
+    addInxVal( &gen_inp.sqlCondInp, COL_META_RESC_ATTR_NAME, tmp_str );
+    addInxIval( &gen_inp.selectInp, COL_META_RESC_ATTR_VALUE, 1 );
+    addInxIval( &gen_inp.selectInp, COL_META_RESC_ATTR_UNITS, 1 );
+    gen_inp.maxRows = 1;
+    status = rsGenQuery( _rsComm, &gen_inp, &gen_out );
+    // if existing entry found, report and exit
+    if ( status >= 0 ) {
+        sqlResult_t* hostname_and_pid;
+        sqlResult_t* timestamp;
+        if ( ( hostname_and_pid = getSqlResultByInx( gen_out, COL_META_RESC_ATTR_VALUE ) ) == nullptr ) {
+            rodsLog( LOG_ERROR, "%s: getSqlResultByInx for COL_META_RESC_ATTR_VALUE failed", __FUNCTION__ );
+            return UNMATCHED_KEY_OR_INDEX;
+        }
+        if ( ( timestamp = getSqlResultByInx( gen_out, COL_META_RESC_ATTR_UNITS ) ) == nullptr ) {
+            rodsLog( LOG_ERROR, "%s: getSqlResultByInx for COL_META_RESC_ATTR_UNITS failed", __FUNCTION__ );
+            return UNMATCHED_KEY_OR_INDEX;
+        }
+        std::stringstream msg;
+        msg << "A rebalance_operation on resource [";
+        msg << _resource_name.c_str();
+        msg << "] is still active (or stale) [";
+        msg << &hostname_and_pid->value[0];
+        msg << "] [";
+        msg << &timestamp->value[0];
+        msg << "]";
+        rodsLog( LOG_ERROR, "%s: %s", __FUNCTION__, msg.str().c_str() );
+        addRErrorMsg( &_rsComm->rError, REBALANCE_ALREADY_ACTIVE_ON_RESOURCE, msg.str().c_str() );
+        return REBALANCE_ALREADY_ACTIVE_ON_RESOURCE;
+    }
+    freeGenQueryOut( &gen_out );
+    clearGenQueryInp( &gen_inp );
+    return 0;
+}
+
+int _set_rebalance_timestamp_avu_on_resource(
+    rsComm_t* _rsComm,
+    const std::string& _resource_name) {
+
+    // get hostname
+    char hostname[MAX_NAME_LEN];
+    gethostname(hostname, MAX_NAME_LEN);
+    // get PID
+    int pid = getpid();
+    // get timestamp in defined format
+    const boost::posix_time::ptime timeUTC = boost::posix_time::second_clock::universal_time();
+    std::stringstream stream;
+    boost::posix_time::time_facet* facet = new boost::posix_time::time_facet();
+    facet->format("%Y%m%dT%H%M%SZ");
+    stream.imbue(std::locale(std::locale::classic(), facet));
+    stream << timeUTC;
+    // add AVU to resource
+    modAVUMetadataInp_t modAVUMetadataInp;
+    memset( &modAVUMetadataInp, 0, sizeof( modAVUMetadataInp ) );
+    modAVUMetadataInp.arg0 = "set";
+    modAVUMetadataInp.arg1 = "-R";
+    modAVUMetadataInp.arg2 = const_cast<char*>(_resource_name.c_str());
```

Comment:
Using `const_cast` makes me nervous.  If `rsModAVUMetadata` ever modifies the content obtained from `_resource_name.c_str()`, then the app will now have undefined behavior.  This can be avoided by copying the data.